cpp/src/fil/infer.cu

```diff
@@ -384,43 +385,47 @@ struct tree_aggregator_t<NITEMS, CATEGORICAL_LEAF> {
     } else {
       finalize_class_label(out, num_rows);
     }
+    __syncthreads();  // free votes[] for next row set
   }
 };
 
 template <int NITEMS, leaf_algo_t leaf_algo, class storage_type>
 __global__ void infer_k(storage_type forest, predict_params params) {
-  // cache the row for all threads to reuse
-  extern __shared__ char smem[];
-  float* sdata = (float*)smem;
-  size_t rid = blockIdx.x * NITEMS;
-  for (int j = 0; j < NITEMS; ++j) {
-    for (int i = threadIdx.x; i < params.num_cols; i += blockDim.x) {
-      size_t row = rid + j;
-      sdata[j * params.num_cols + i] =
-        row < params.num_rows ? params.data[row * params.num_cols + i] : 0.0f;
+  for (int grid_row0 = 0; grid_row0 < params.num_rows;
```

Comment:
`for (int start_row = blockIdx.x * NITEMS; start_row < params.num_rows; start_row += NITEMS * params.num_blocks)`

Reply:
why `start_row`?

Reply:
Feel free to use a different name. The idea is that you get the value of `block_row0` directly, and don't need an extra `grid_row0` variable.